iscsid: Rate limit session reopen log messages

gulams opened the pull request:
When the iscsi targets disappear due to any reason like ports down or the network issues or any other reason, existing sessions will be retried by the iscsi daemon and each retry will have a 3 seconds gap by default. So, the session reopen logs are logged into /var/log/messages for every 3 seconds and due to this the messages file is blown out. To avoid this blow out of messages file, a config parameter is provided in iscsid.conf file in this feature using which a user can tune the interval between the subsequent session reopen log messages By default the value of this config option in iscsid.conf file (node.session.reopen_log_freq) is 1 and is commented out currently.
If user wants to change this value, he can uncomment this option and set the value to a desirable value like 2 or 3 or 4 and restart iscsid. If the value for this config value is 4, then these session reopen logs will be logged for every 12 seconds (4 * 3 seconds = 12 seconds). This feature works only if the default log level in iscsid.service file is set to <= 2

gonzoleeman commented on etc/iscsid.conf:
```diff
@@ -347,6 +347,18 @@ node.session.nr_sessions = 1
+# reopen logs will be logged for every 12 seconds (4 * 3 seconds = 12
+# seconds).
+# node.session.reopen_log_freq = 1
+
```
Many modern linux systems (like tumbleweed) do not use /var/log/messages, but instead use the systemd journal. Perhaps you can word your description here more generically, something like "session reopen logging occurs every 3 seconds, and can overwhelm the logging system>"?

gonzoleeman commented on libopeniscsiusr/default.h:
```diff
@@ -57,6 +57,9 @@
 /* session reopen max retries */
 #define	DEF_SESSION_REOPEN_MAX		0
 
+/* session reopen log frequency */
```
could you add something to the comment like "number of reopens per message", since "frequency" could be ambiguously interpreted?

gulams replied:
Done

gonzoleeman commented on libopeniscsiusr/idbm.h:
```diff
@@ -220,6 +220,7 @@ struct iscsi_session_idbm {
 	char					boot_root[BOOT_NAME_MAXLEN];
 	char					boot_nic[BOOT_NAME_MAXLEN];
 	char					boot_target[BOOT_NAME_MAXLEN];
+	int64_t                                 reopen_log_freq;
```
Do you really need 64 bits here for a number that will generally fairly small?

gulams replied:
Yes, this needs int64_t as _idbm_node_rec_link() in libopeniscsiusr/idbm.c calls _rec_int64() which needs int64_t.

gonzoleeman commented on usr/config.h:
```diff
@@ -199,6 +199,7 @@ typedef struct session_rec {
 	char					boot_root[BOOT_NAME_MAXLEN];
 	char					boot_nic[BOOT_NAME_MAXLEN];
 	char					boot_target[BOOT_NAME_MAXLEN];
+	int                                     reopen_log_freq;
```
ok, it's int64_t above (which seemed excessive). Since you're using int here, how about int above?

gulams replied:
The above needs int64_t as mentioned by me in the above comment. Here it needs int as the function idbm_recinfo_node() in usr/idbm.c calls __recinfo_int() which needs int.

gonzoleeman commented on usr/discovery.c:
```diff
@@ -823,6 +823,7 @@ iscsi_alloc_session(struct iscsi_sendtargets_config *config,
 	session->conn[0].noop_out_timeout = 0;
 	session->conn[0].noop_out_interval = 0;
 	session->reopen_cnt = config->reopen_max + 1;
+	session->reopen_log_freq = 1;
```
It's strange that we have default value defines for these things but don't use them here ... But it's not your fault that the usr-vs-libopeniscsiusr database stuff is a mess of duplication. An contributor did that, and I should have caught it before merging his code.

gulams replied:
Changed to defaults.

gonzoleeman commented on usr/event_poll.c:
```diff
@@ -214,7 +214,7 @@ void event_loop(struct iscsi_ipc *ipc, int control_fd, int mgmt_ipc_fd)
 					break;
 				} else {
-					log_debug(1, "Poll was woken by an alarm");
+					log_debug(4, "Poll was woken by an alarm");
```
So that's at least two debug log messages that have changed the level at which they print, i.e. to only print at a higher debug level. That's fine, but I think that should be mentioned in the description, as long as you are updating it (since it needs to be more clear about the "log level" being the "-d" debug log level of iscsid.

gonzoleeman commented on usr/initiator.c:
```diff
@@ -260,6 +260,9 @@ __session_conn_create(iscsi_session_t *session, int cid)
 	session->reopen_max = rec->session.reopen_max;
+	session->reopen_log_freq = rec->session.reopen_log_freq;
+	if (session->reopen_log_freq == 0)
+		session->reopen_log_freq = 1;
```
Perhaps you should just set it to "1" if it is less than 1? So we can handle negative numbers (and ignore them)?

Also, you should mention in iscsid.conf that numbers below 1 will be treated as "1". Lastly, seems like there should be a "maximum"? What if they set it to 10,000? Isn't that really "just turn it off"? Seems like, at the least, some maximum like "100" would be reasonable. And perhaps we should allow "0" to mean "just turn it off"? What do you think?

gonzoleeman commented on usr/log.c:
```diff
@@ -336,6 +336,20 @@ void log_info(const char *fmt, ...)
 	va_end(ap);
 }
 
+void log_session_reopen(int level, iscsi_session_t *session, const char *fmt, ...)
```
I dislike that we need a new function just for this one case. Maybe it should be called something more generic, in case we want to use it elswhere, like "log_debug_limited()" or "log_debug_freq_limit()" I'd rather have a long name than an obscure one. But it needs "debug" in the title, since normally debug logging is set to 0 (none) for many installations.

It looks like 6 messages (if I count correctly) have changed from log_debug() to log_session_reopen(). How did you pick those? No, I see one was a "log_error()".  Isn't that the one that really started all this, i.e. log_error("connect to %s:%s failed (%s)");.

gonzoleeman commented on usr/log.c:
```diff
@@ -336,6 +336,20 @@ void log_info(const char *fmt, ...)
+	int reopen_log_freq = session->reopen_log_freq;
+
+	if ((log_level >= level) && ((log_level > 2) ||
+		(reopen_cnt % reopen_log_freq) == 0)) {
```
so this prints a message every time, even if debug logging isn't enabled, if the freq is 1? And if it's say, above 1, like e.g. 10, you would get a message every tenth time, plus every other time if debug logging is enabled?

That makes my head hurt ...

gulams replied:
Yes, you are right. Changed to print only if the debug logging is enabled.

gonzoleeman commented on usr/log.c:
```diff
@@ -336,6 +336,20 @@ void log_info(const char *fmt, ...)
+		(reopen_cnt % reopen_log_freq) == 0)) {
+			va_list ap;
+			va_start(ap, fmt);
+			log_func(LOG_DEBUG, log_func_priv, fmt, ap);
```
but you're still logging at the debug level, even if debug logging level is set to 0 in iscsid?

gonzoleeman commented on usr/initiator.c:
```diff
@@ -588,7 +591,7 @@ __session_conn_reopen(iscsi_conn_t *conn, queue_task_t *qtask, int do_stop,
 	uint32_t delay;
 
-	log_debug(1, "re-opening session %d (reopen_cnt %d)", session->id,
+	log_session_reopen(1, session, "re-opening session %d (reopen_cnt %d)", session->id,
```
so instead of logging this message if debug logging is enabled at all, now it prints it every time, by default (since freq=1 by default)?

gulams replied:
Changed.

gulams commented:
> Hi @gulams . thank you for this submission. My apologies for not looking at it sooner.
> 
> I've just started reading, and I have a question about the description. What do you mean by: "This feature works only if the default log level in iscsid.service file is set to <= 2" ???

When the log level in iscsid.service file is set above 2, then we don't want to rate limit the session re-open logs as the users usually want all the logs to be present in the messages file.

gonzoleeman reviewed (changes requested):
Thank you for submitting this. I appreciate the work it takes.

I have no issue with the idea of limiting the error messages when retries occur. I'm not sure I like some of the implementation details. I'd like to hear @mikechristie 's thoughts ...

As mentioned in my code comments:
- I think the description in iscsid.conf need to have a bit more
- Perhaps we should allow 0 to mean "no messages", and limit the max frequency? Certainly we need to document the real limits of this "freq' attribute.
- I don't like the new logging function name. The function seems to mix the "freq" value with the debug log level. That needs to be cleaned up. I'd like to see it be more generic.
- You changed some messages from log_debug() to your new function. Not sure why. To limit their frequency? The need for that, and the effect it has, should be documented as part of this.

Let me know what you think about these issues. I'd like to work with you to ensure this pull request succeeds. Thanks again for your efforts!

mikechristie commented on etc/iscsid.conf:
```diff
@@ -347,6 +347,18 @@ node.session.nr_sessions = 1
 
+# When the iscsi targets disappear due to any reason like ports down or
+# the network issues or any other reason, existing sessions will be retried
+# by the iscsi daemon and each retry will have a 3 seconds gap by default.
```
We use defaultTime2Wait we get from the target during login negotiation for connection time outs. For pretty much everything else we use that hard coded 3 seconds.

mikechristie commented on usr/actor.c:
```diff
@@ -67,7 +67,7 @@ actor_delete(actor_t *thread)
 		/* fallthrough */
 	case ACTOR_SCHEDULED:
-		log_debug(1, "deleting a scheduled/waiting thread %p:%s!",
+		log_debug(4, "deleting a scheduled/waiting thread %p:%s!",
```
I would put this in a different patch and note why it changed. I think Chris or someone (maybe me) made it a 1 because it shouldn't happen often. However, I think something changed and we see it all the time now.

gulams commented:
> I think the root issue is that I messed up the logging originally :) Right now, for the debug logging we have a severity level but sometimes ... a lot of times ... the values are not consistent. I probably chose some based on whatever I felt I wanted to see that day. For example, the connection related log_debug 1 messages you are changing were important to me because I tested that a lot. The actor one was probably important for Chris who I think fixed that code. When RHEL/OL turns on the logging debugging now in some versions by default these messages can be annoying/overkill.
> 
> I guess I'm saying there are 2 issues
> 
>     1. We should categorize the messages. For example this patch starts this by putting the connection related debug messages into a group.
> 
>     2. Add rate limiting like you did.
> 
> 
> In the end maybe the goal could be the user can specific a TYPE (LOGIN, CONNECT, RETRYABLE_ERR, KERNEL, SYSFS, IDBM, EVENT, ACTOR, etc) and a frequency?
> 
> Do you think this is overkill? I'm not sure how useful it will be. Do users really just care about the login/connect related messages?

The issue which our users encountered is that, when an interface goes down, the iscsi connections will go down and the iscsi will retry for recovery of those sessions till they are restored. But if the interface is down forever, then the sessions will not be recovered and the logging system will be flooded with a lot of session reopen messages filling up the messages file. The users are concerned that, due to this overwhelming of the logging system, they may lose the  important diagnostics for other problems which may lead to other issues for them.

To avoid this losing of other important diagnostics for other issues, they wanted these session recovery messages to be rate limited so that the diagnostics for other important problems are retained. 

Other than this, we haven't seen any users reported any concern about other messages being logged from iscsi. They are only concerned about these session recovery messages as they are logged very frequently.

I am not sure if the categorization of the messages, as mentioned by you, is beneficial as we have seen only the above concern from the users. If it has any other benefits, can you please let us know so that we can get good understanding?
Can you please suggest?

gonzoleeman commented:
Apologies for not following up more quickly here. I will try to look over your updates soon and approve your request if possible. Thank you for your submission and your patience. Have a great holidays!

gulams commented:
@gonzoleeman, Can you please review my changes above and provide feedback?

gonzoleeman commented:
Apologies for my tardiness. I will review as soon as I can. --Lee-Man DuncanSent from my iPhone, dudeOn Jan 15, 2025, at 2:14 PM, Gulam Mohamed ***@***.***> wrote:﻿
@gonzoleeman, Can you please review my changes above and provide feedback?

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

gonzoleeman reviewed (dismissed):
All looks good now.

gulams commented:
> All looks good now.

Thanks Lee for the review and approval. Just want to know when it will be merged

gonzoleeman commented:
Ok, the code looks excellent now. Thank you for all your work.

I have only one more request. Please replace the commit message for your very last commit, as:

>     Comments addressed

Is a little sparse. Yes, you addressed the comments, but what does that mean?

You:

- clarified some comments
- raised the log level of a debug log message (and why you did that)
- renamed the logging function to be more generic
- some cosmetic cleanup

For example.

gonzoleeman commented:
I've resolved the issue I brought up.

@mikechristie Thank you for your reviews! Are you happy with the resolution of the issue you brought up?

gonzoleeman commented:
On Jan 21, 2025, at 10:43 AM, Gulam Mohamed ***@***.***> wrote:
> 
> 
> All looks good now.
> 
> Thanks Lee for the review and approval. Just want to know when it will be merged
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/open-iscsi/open-iscsi/pull/492#issuecomment-2605485016>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAOJGHFIXSPTNOWGXWNJAQD2L2IL3AVCNFSM6AAAAABSNJ464WVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMMBVGQ4DKMBRGY>.
> You are receiving this because you were mentioned.
> 

I would like you to change the final commit message, if you would. You can just force push your update once you do that.

I believe all of Mike’s issues were addressed too, but I’m going to give him a day or to so that he can respond on his own, if he likes.

So, all things considered, if Mike has no more requests, it should be later this week.

I may tag a new version after that, since it’s been a while, and tagging a new version makes it a bit more likely that other distros will pick up the changes sooner rather than later.

Thank you again for diligence. :)
— 
Lee

gulams commented:
> Ok, the code looks excellent now. Thank you for all your work.
> 
> I have only one more request. Please replace the commit message for your very last commit, as:
> 
> > ```
> > Comments addressed
> > ```
> 
> Is a little sparse. Yes, you addressed the comments, but what does that mean?
> 
> You:
> 
>     * clarified some comments
> 
>     * raised the log level of a debug log message (and why you did that)
> 
>     * renamed the logging function to be more generic
> 
>     * some cosmetic cleanup
> 
> 
> For example.

Thanks Lee for your help on this.
As per your suggestion, changed the last commit message.

Thanks again for your help.
Regards,
Gulam Mohamed

gonzoleeman commented:
I have reviewed your latest update. Actually, I mostly reviewed your code after the last update, compared to "master".

I am going to approve your request, but I wanted to point out some issues, so that if you make another request like this, it can go more easily.

In general, we try to mostly follow the kernel git guidelines, but we certainly aren't as strict as the kernel folks rightly are.

Here are some comments on your submission:

- Generally, when you have comments on  your code that result in changes, you resubmit, rather than just add another commit on top of your current submission. The reason for this is that upstream does not need to capture the intermediate steps that lead to your final submission. NOTE: A complicated merge request might justify multiple commits to be merged, but usually a "one commit, one issue" approach is good.
- Two of your three commits have a first line that ends in a colon, e.g. "44852d9fb4dd Addressed the following comments:". But the first line of a commit message is a summary. A summary more like "Addressed review comments" would be better. But even better would be to resubmit the code after changing as requested. (If you need advice on how to resubmit a branch, using "git push --force", please let me know.)
- I still don't really like the function name, nor it's function. :) Even though its named better now (conn_log_connect), it doesn't reflect the fact that it only prints in **debug** log mode. The function it's replacing was called "log_debug", which made it clear that it was for debugging messages. We also have "log_info()", and "log_error()", for messages that are not debugging messages.

But as Mike pointed out, the logging system is a little long in the tooth (i.e. old), and really should be improved. Something I might attempt some time, but like you, not now.

gonzoleeman commented:
Okay, one last issue. You did not do as I Mike requested and put the log_debug() change in actor.c in a separate patch (you changed it from debug level 1 to level 4). I've left that request of his as "unresolved". (See above)

Given this, I'd appreciate it if you squashed all your changes into two commits:
- One for everything but the actor.c/log_debug() change mentioned, and
- One more for the actor.c/log_debug() change

You can use your original (first commit) commit message, with changes as needed, if any, based on the code changes you did.

Let me reiterate that I want to merge this, so please don't give up now. Thanks for your patience.
